pkg/providers/instancetype/offering/provider.go

```diff
@@ -0,0 +1,169 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package offering
+
+import (
+	"context"
+	"fmt"
+
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+
+	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	awscache "github.com/aws/karpenter-provider-aws/pkg/cache"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/capacityreservation"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/pricing"
+)
+
+type Provider interface {
+	InjectOfferings(context.Context, []*cloudprovider.InstanceType, *v1.EC2NodeClass, []string) []*cloudprovider.InstanceType
+}
+
+type DefaultProvider struct {
+	unavailableOfferings        *awscache.UnavailableOfferings
+	pricingProvider             pricing.Provider
+	capacityReservationProvider capacityreservation.Provider
+}
+
+func NewDefaultProvider(unavailableOfferingsCache *awscache.UnavailableOfferings, pricingProvider pricing.Provider) *DefaultProvider {
+	return &DefaultProvider{
+		unavailableOfferings: unavailableOfferingsCache,
+		pricingProvider:      pricingProvider,
+	}
+}
+
+func (p *DefaultProvider) InjectOfferings(
+	instanceTypes []*cloudprovider.InstanceType,
+	nodeClass *v1.EC2NodeClass,
+	allZones sets.Set[string],
+) []*cloudprovider.InstanceType {
+	subnetZones := lo.SliceToMap(nodeClass.Status.Subnets, func(s v1.Subnet) (string, string) {
+		return s.Zone, s.ZoneID
+	})
+	its := []*cloudprovider.InstanceType{}
+	for _, it := range instanceTypes {
+		offerings := p.createOfferings(
+			it,
+			nodeClass,
+			allZones,
+			subnetZones,
+		)
+		for _, of := range offerings {
+			InstanceTypeOfferingAvailable.Set(float64(lo.Ternary(of.Available, 1, 0)), map[string]string{
+				instanceTypeLabel: it.Name,
+				capacityTypeLabel: of.Requirements.Get(karpv1.CapacityTypeLabelKey).Any(),
```

Comment:
This is a small thing but how does this availability look for reserved since they can go in and out of available depending on the last offering that you get -- there's a larger but with this metric [here](https://github.com/aws/karpenter-provider-aws/issues/7758) so I'm also fine if you want to punt it out

Reply:
That's a good point, I think I see a couple of options:
- Mark it as available if any reserved offering is available
- Expand the label set to include a reservation id label

Either one would be pretty easy so I'm happy to implement it in this PR. I'm leaning towards the first option, since other offerings obviously don't have associated reservation IDs. I'm also thinking down the road we should implement a metric to for reservation availability, so I don't think there's a need to encode individual reservation availability here.

Reply:
Discussed offline: I think we're both in agreement that "Mark it as available if any reserved offering is available" is the best way to do this.